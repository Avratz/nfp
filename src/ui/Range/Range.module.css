.rangeContainer {
  font-family: var(--font-poppins);
  font-weight: var(--font-medium);
  text-transform: uppercase;
  font-size: var(--text-sm);
}

.label {
  color: var(--neutral-5);
  font-size: var(--text-xs);
}

.units {
  display: flex;
  justify-content: space-between;
  color: var(--neutral-8);
  margin-top: 0.25rem;
}

.inputRange {
  width: 100%;
  height: 24px;
  appearance: none;
  cursor: pointer;
  margin: 0;
  position: absolute;
  z-index: 15;
  background: transparent;
}

.inputRange::-webkit-slider-runnable-track {
  height: 24px;
}

.inputRange::-moz-range-track {
  height: 24px;
}

.inputRange::-webkit-slider-thumb {
  appearance: none;
  height: 100%;
  height: 20px;
  width: 20px;
}

.rangeBox {
  position: relative;
  width: 100%;
  height: 24px;
}

.track,
.progress {
  width: 100%;
  height: 20px;
  clip-path: polygon(100% 0%, 100% 100%, 0% 55%, 0% 45%);
  border-radius: 16px;
  position: absolute;
  top: 50%;
  transform: translateY(-50%);
}

.track {
  z-index: 10;
}

.progress {
  height: 16px;
  z-index: 11;
}

.dot {
  background-color: var(--accent);
  height: 20px;
  width: 20px;
  border: 4px solid var(--neutral-8);
  border-radius: 50%;
  position: absolute;
  top: 2px;
  z-index: 12;
}

.tooltip {
  position: absolute;
  top: -45px;
  left: 50%;
  transform: translateX(-50%);
  width: 70px;
  height: 30px;
  border-radius: 0.5rem;
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: var(--neutral-8);
  color: var(--neutral-2);
  font-size: var(--text-xs);
}

.tooltip::after {
  content: '';
  display: block;
  height: 20px;
  width: 20px;
  background-color: var(--neutral-8);
  position: absolute;
  bottom: -5px;
  left: calc(50% - 10px);
  clip-path: polygon(0% 0%, 100% 100%, 0% 100%);
  transform: rotate(-45deg);
  border-radius: 0 0 0 0.4rem;
}
