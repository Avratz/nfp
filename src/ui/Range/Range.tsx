import { ChangeEvent, InputHTMLAttributes, useEffect, useState } from 'react'
import styles from './Range.module.css'

interface RangeProps extends InputHTMLAttributes<HTMLInputElement> {
  label?: string
  unit: string
  value?: number
  min: number
  max: number
}

export function Range({
  label,
  unit,
  onInput,
  value,
  min,
  max,
  ...props
}: RangeProps) {
  const [selectedValue, setSelectedValue] = useState<number>(value || max / 2)

  function handleInput(e: ChangeEvent<HTMLInputElement>) {
    let value = Number(e.currentTarget.value)
    setSelectedValue(value)

    if (onInput) {
      onInput(e)
    }
  }

  return (
    <label className={styles.rangeContainer}>
      <div className={styles.label}>{label ? label : null}</div>
      <div className={styles.rangeBox}>
        <input
          type="range"
          className={styles.inputRange}
          onInput={handleInput}
          value={value}
          min={min}
          max={max}
          {...props}
        />
        <div className={styles.track}></div>
        <div
          className={styles.progress}
          style={{
            background: calculateLinearGradient(
              calculatePercentage(selectedValue, max),
            ),
          }}
        ></div>
        <div
          className={styles.dot}
          style={{
            left: calculatePercentage(selectedValue, max),
            transform: `translateX(-${calculatePercentage(
              selectedValue,
              max,
            )})`,
          }}
        >
          <div className={styles.tooltip}>{selectedValue} ETH</div>
        </div>
      </div>

      <datalist className={styles.units}>
        <option label={`${min} ${unit}`} />
        <option label={`${max} ${unit}`} />
      </datalist>
    </label>
  )
}

function calculatePercentage(value: number, total: number) {
  let percentage = (value * 100) / total ?? 50
  return percentage + '%'
}

function calculateLinearGradient(percentage: string) {
  return `linear-gradient(to right, var(--accent) 0%, var(--accent) ${percentage}, var(--neutral-3) ${percentage}, var(--neutral-3) 100%)`
}
