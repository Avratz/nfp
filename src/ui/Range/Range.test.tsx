import { render } from '@testing-library/react'
import { Range } from './Range'

describe('Range', () => {
  it('renders default', () => {
    const { container } = render(<Range unit="ETH" min={5} max={10} />)
    expect(container).toMatchSnapshot()
  })

  it('renders with label', () => {
    const { container } = render(
      <Range label="Range" unit="ETH" min={5} max={10} />,
    )
    expect(container).toMatchSnapshot()
  })
})
