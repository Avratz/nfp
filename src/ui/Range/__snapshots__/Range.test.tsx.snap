// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Range renders default 1`] = `
<div>
  <label
    class="rangeContainer"
  >
    <div
      class="label"
    />
    <div
      class="rangeBox"
    >
      <input
        class="inputRange"
        max="10"
        min="5"
        type="range"
        value=""
      />
      <div
        class="track"
      />
      <div
        class="progress"
      />
      <div
        class="dot"
        style="left: 50%; transform: translateX(-50%);"
      >
        <div
          class="tooltip"
        >
          5
           ETH
        </div>
      </div>
    </div>
    <datalist
      class="units"
    >
      <option
        label="5 ETH"
      />
      <option
        label="10 ETH"
      />
    </datalist>
  </label>
</div>
`;

exports[`Range renders with label 1`] = `
<div>
  <label
    class="rangeContainer"
  >
    <div
      class="label"
    >
      Range
    </div>
    <div
      class="rangeBox"
    >
      <input
        class="inputRange"
        max="10"
        min="5"
        type="range"
        value=""
      />
      <div
        class="track"
      />
      <div
        class="progress"
      />
      <div
        class="dot"
        style="left: 50%; transform: translateX(-50%);"
      >
        <div
          class="tooltip"
        >
          5
           ETH
        </div>
      </div>
    </div>
    <datalist
      class="units"
    >
      <option
        label="5 ETH"
      />
      <option
        label="10 ETH"
      />
    </datalist>
  </label>
</div>
`;
